runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,297 @@
+  local row_level = {}
+  ---@type table<integer, string>
+  local row_text = {}
+  ---@type table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+  local kind_ranges = {}
+
+  for _, range in ipairs(result) do
+    local kind = range.kind
+    if kind then
+      ---@cast kind lsp.FoldingRangeKind
+      local ranges = kind_ranges[kind] or {}
+      ranges[#ranges + 1] = range
+      kind_ranges[kind] = ranges
+    end
+
+    local start_row = range.startLine
+    local end_row = range.endLine
+    -- Fold a single line is not supported
+    if start_row ~= end_row then
+      row_text[start_row] = range.collapsedText
+      for row = start_row, end_row do
+        local level = row_level[row] or { 0 }
+        level[1] = level[1] + 1
+        row_level[row] = level
+      end
+      row_level[start_row][2] = '>'
+      row_level[end_row][2] = '<'
+    end
+  end
+
+  bufstate.row_level = row_level
+  bufstate.row_text = row_text
+  bufstate.kind_ranges = kind_ranges
+  bufstate.version = ctx.version
+
+  if api.nvim_get_mode().mode:match('^i') then
+    -- `foldUpdate()` is guarded in insert mode.
+    schedule_foldupdate(bufnr)
+  else
+    foldupdate(bufnr)
+  end
+end
+
+local augroup_refresh = vim.api.nvim_create_augroup('vim_lsp_folding_range/refresh', {})
+
+---@param bufnr integer
+---@param client_id integer The ID of the client whose server supports `foldingRangeProvider`.
+---@return integer? request_id
+local function refresh(bufnr, client_id)
+  ---@type lsp.FoldingRangeParams
+  local params = { textDocument = util.make_text_document_params(bufnr) }
+  local client = assert(vim.lsp.get_client_by_id(client_id))
+  local _, request_id = client:request(ms.textDocument_foldingRange, params, nil, bufnr)
+  return request_id
+end
+
+--- Setup event hooks for folding range.
+---@param bufnr integer
+---@return vim.lsp.folding_range.BufState
+local function setup(bufnr)
+  api.nvim_create_autocmd('LspNotify', {
+    group = augroup_refresh,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      if
+        client.server_capabilities.foldingRangeProvider
+        and (
+          args.data.method == ms.textDocument_didChange
+          or args.data.method == ms.textDocument_didOpen
+        )
+      then
+        refresh(bufnr, args.data.client_id)
+      end
+    end,
+  })
+
+  ---@type vim.lsp.folding_range.BufState
+  local bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+  bufstates[bufnr] = bufstate
+
+  api.nvim_buf_attach(bufnr, false, {
+    on_detach = function()
+      bufstates[bufnr] = nil
+    end,
+    on_reload = function()
+      bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+      bufstates[bufnr] = bufstate
+      for _, client in
+        ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+      do
+        if client.server_capabilities.foldingRangeProvider then
```

Comment:
Oh, I know the reason now, `foldingRange` was not added to the supported methods of `client:supports_method` 😆. These checks should not be necessary because there is `get_clients()`